File: gbControlUDPServerDev/external_comm.py
import socket
import time
import threading
import os


# will send packets with added layers of directional transmitting
# a packet will have a file based layer and a function based layer
# the packet has the file layer wrapped around the base layer, aswell as the base layer wrapping the core data
# all transmission layers are a single byte long

class E_COMM:
    def __init__(self, target_address:tuple,id:str, is_robot:bool) -> None:
        self.server_address = target_address
        self.connect_id = ("<" + id + ">").encode()

        self.is_robot = is_robot

        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        #self.sock.bind(("localhost", port))
        self.sock.settimeout(0.1) # timeout after 0.1 seconds
        self.buffer_size = 65536

        self.latency_millis = 0.
        self.last_receive_time = 0.
        self.connection_status = False

        self.data_prefix = {'connect_request':'\x00'.encode(), 
                            'connection_check':'\x01'.encode(), 
                            'ack':'\x02'.encode(), 
                            'data':'\x03'.encode(), 
                            'punch_hole':'\x04'.encode(), 
                            'address':'\x05'.encode(),
                            "client":"\x00".encode(),
                            "robot":"\x01".encode()
                        }


    def _send(self, data:bytes, address:tuple):    
        if len(data) > self.buffer_size or address == (): 
            return 0 # data too large, doesnt fit in packet
        try:
            #print(data, address)
            self.sock.sendto(
                (self.data_prefix["client"] + self.data_prefix["robot"]) if self.is_robot else (self.data_prefix["robot"] + self.data_prefix["client"])
                 + data, address)
        except socket.gaierror: # means that there is no correct address to send to 
            print("incorrect address")
            return 0
        except socket.timeout:
            return 0
        #except:
         #   print("unknown error")
          #  pass
            # os._exit(1)
        return 1
    
    def send(self, data:bytes): # will send data to connected client
        return self._send(self.connect_id + data, self.server_address)
    
    def send_s(self, data:bytes):
        self.send(self.data_prefix['data'] + data)
        return 1
    
    def receive_loop(self, recv_function): # will run forever and requires a function to be passed that handles data throughput 
        # DO NOT PASS A BLOCKING FUNCTION, will not allow for functional communication
        # every time interval send a "connection check" 
        # make sure that a response is received every time interval + extra time
        last_connection_check_time = 0.
        connection_check_interval = 0.5

        last_ack_time = 0
        max_ack_time = 1.

        while True:

            if time.time() - last_connection_check_time > connection_check_interval:
                self.send(self.data_prefix['connection_check'] + str(time.time()).encode())
                last_connection_check_time = time.time()
            self.connection_status = time.time() - last_ack_time < max_ack_time

            try:
                data, address = self.sock.recvfrom(self.buffer_size)
            except socket.timeout:
                continue # nothing to be done just wait for next loop
            except ConnectionResetError:
                continue
            
            
            data_target = None


            if address == self.server_address:
                try:
                    decode_data = data[:len(self.connect_id)].decode()
                except: 
                    continue
                if self.connect_id == data[decode_data.find('<'):decode_data.find('>')+1]:
                    data = data[decode_data.find('>')+1:]
                    data_target = data[:1]
                    data = data[1:]
                else: continue
                # first layer of prefix is unwrapped here
                #print("received data: ", data)
                if data_target == self.data_prefix['data']:
                        recv_function(data) # run recv function with specified data
                        self.last_receive_time = time.time()
                        continue
                
                # connection dependent on received data
                if data_target == self.data_prefix['connection_check']:
                    self.send(self.data_prefix['ack'] + data)
                if data_target == self.data_prefix['ack']:
                    last_ack_time = time.time()
                    try:
                        self.latency_millis = (time.time() - float(data)) * 1000
                    except ValueError:
                        # could not convert float (data)
                        pass

File: gbControlLite/src/utils/externalComm.ts
// import DgramAsPromised from "dgram-as-promised"
import { Buffer } from 'buffer';
import { UDP } from '@nws/capacitor-udp'

export class EComm {
    private serverAddress: [string, number];
    private connectId: Buffer;
    private isRobot: boolean;
    private sock: any;
    private id: string;
    private sockID: number;
    private bufferSize: number;
    private latencyMillis: number;
    private lastReceiveTime: number;
    private connectionStatus: boolean;
    private dataPrefix: { [key: string]: Buffer };

    constructor(targetAddress: [string, number], id: string, isRobot: boolean) {
        this.serverAddress = targetAddress;
        this.connectId = Buffer.from(`<${id}>`);
        // this.isRobot = isRobot;
        this.isRobot = false; // This is the client app
        this.sockID = 1; // Initialize to 1
        this.id = id;
        
        // UDP.joinGroup({socketId: this.sockID, address: this.serverAddress[0]});
        // this.sock.setTimeout(100); // timeout after 0.1 seconds
        this.bufferSize = 65536;
        this.latencyMillis = 0;
        this.lastReceiveTime = 0;
        this.connectionStatus = false;
        this.dataPrefix = {
            'connect_request': Buffer.from('\x00'),
            'connection_check': Buffer.from('\x01'),
            'ack': Buffer.from('\x02'),
            'data': Buffer.from('\x03'),
            'punch_hole': Buffer.from('\x04'),
            'address': Buffer.from('\x05'),
            'client': Buffer.from('\x00'),
            'robot': Buffer.from('\x01')
        };
    }

    private _send(data: Buffer, address: [string, number]): number {
        // console.log("Sending data 123:", data);
        if (data.length > this.bufferSize || address.length === 0) {
            console.warn("Data too large or incorrect address");
            return 0; // data too large, doesn't fit in packet
        }
        try {
            const prefix = this.isRobot ? Buffer.concat([this.dataPrefix["robot"], this.dataPrefix["client"]]) : Buffer.concat([this.dataPrefix["client"], this.dataPrefix["robot"]]);
            // this.sock.send(Buffer.concat([prefix, data]), address[1], address[0]);
            const concatenatedBuffer = Buffer.concat([prefix, data]);
            // console.log("the concatenated buffer:", concatenatedBuffer);
            const bufferString = concatenatedBuffer.toString('base64');
            // console.log("Sending data 123:", bufferString);
            UDP.send({socketId: this.sockID, address: address[0], port: address[1], buffer: bufferString});
        } catch (error) {
            if (error.code === 'EAI_AGAIN') {
                console.error("incorrect address");
            }
            return 0;
        }
        return 1;
    }

    public send(data: Buffer): number {
        // console.log("Sending data BBBB:", data);
        // console.log('Returning data to string BBBB:', data.toString());
        return this._send(Buffer.concat([this.connectId, data]), this.serverAddress);
    }

    public sendS(data: Buffer): number {
        
        this.send(Buffer.concat([this.dataPrefix['data'], data]));
        return 1;
    }

    public receiveLoop(recvFunction: (data: Buffer) => void): void {
        // console.log("Starting receive loop");
        let lastConnectionCheckTime = 0;
        const connectionCheckInterval = 0.5;
        let lastAckTime = 0;
        const maxAckTime = 1;
        UDP.addListener('receive', (data:any) => {
            // console.log("Received data, data:", data);
            const data1 = Buffer.from(data.buffer, 'base64');
            const rinfo = {address: this.serverAddress[0], port: this.serverAddress[1]};
            if ((rinfo.address === this.serverAddress[0] && rinfo.port === this.serverAddress[1])) {
                // console.log("Made it to decode data");
                const decodeData = data1.slice(0, this.connectId.length).toString();
                // console.log("Decoded data:", decodeData);
                if (this.connectId.equals(data1.slice(decodeData.indexOf('<'), decodeData.indexOf('>') + 1)) || true) {
                    // console.log("Made it to step 2");

                    let data2 = data1.slice(decodeData.indexOf('>') + 1);
                    const dataTarget = data2.slice(0, 1);
                    data2 = data2.slice(1);

                    // console.log("Data target:", dataTarget, "Data Prefix:", this.dataPrefix['data']);

                    if (dataTarget.equals(this.dataPrefix['data'])) {
                        recvFunction(data2);
                        this.lastReceiveTime = Date.now();
                    } else if (dataTarget.equals(this.dataPrefix['connection_check'])) {
                        this.send(Buffer.concat([this.dataPrefix['ack'], data1]));
                    } else if (dataTarget.equals(this.dataPrefix['ack'])) {
                        lastAckTime = Date.now() / 1000;
                        try {
                            this.latencyMillis = (Date.now() / 1000 - parseFloat(data.toString())) * 1000;
                        } catch (error) {
                            // could not convert float (data)
                        }
                    }
                }
            }
        });

        UDP.addListener('receiveError', (error: any) => {
            console.error("Error receiving data: ", error);
        });

        setInterval(() => {
            if (Date.now() / 1000 - lastConnectionCheckTime > connectionCheckInterval) {
                this.send(Buffer.concat([this.dataPrefix['connection_check'], Buffer.from(Date.now().toString())]));
                lastConnectionCheckTime = Date.now() / 1000;
            }
            this.connectionStatus = Date.now() / 1000 - lastAckTime < maxAckTime;

            
        }, 500);
    }

    public async initialize(): Promise<number> {
        this.sock = UDP.create({properties: {name: this.id, bufferSize:65536}});
        // console.log("The address port is:", this.serverAddress[1]);
        await UDP.bind({socketId: this.sockID, port: this.serverAddress[1]});
        
        // console.log("Initializing");
        return 1;
    }
}